lib/Sema/CSApply.cpp

```diff
@@ -272,6 +272,20 @@ void ConstraintSystem::propagateLValueAccessKind(Expr *E,
                                allowOverwrite);
 }
 
+bool ConstraintSystem::isTypeReference(Expr *E) {
+  return E->isTypeReference([&](const Expr &E) -> Type { return getType(&E); });
```

Comment:
I recall what the issue is now. I believe you end up having to use std::bind and potentially also placeholders, and the whole mess isn't much less complicated than the explicit lambda.